### ISSUE TITLE
Doxygen regression when creating PDF output 1.9.6 -> master of 10.01.2023

### ISSUE THREAD

**Issue by StefanBattmer** [doxygen/doxygen#9781]
**Describe the bug**
The PDF generation is no longer generating an output but stops with an error message now:

```
! Undefined control sequence.
l.3 ...of the manual, you will get an \doxysectref
                                                  {introduction}{p.}{mvBF_pa...

? 
! Emergency stop.
l.3 ...of the manual, you will get an \doxysectref
                                                  {introduction}{p.}{mvBF_pa...

!  ==> Fatal error occurred, no output PDF file produced!
```

**Expected behavior**
A correctly generated PDF document

**Screenshots**
none

**To Reproduce**
Could try to come up with something when the attached snippets don't ring a bell but I am afraid that this might turn out difficult this time...

**Version**
1.9.6 is still working fine, The master from today (10.01.2023) is not

**Stack trace**
Have attached the 2 relevant sections from the output instead for diffing. Hope that helps.

[DoxygenIssue9781.zip](https://github.com/doxygen/doxygen/files/10381631/DoxygenIssue9781.zip)

**Comment by albert-github** [comment]
From the 2 text files we see that in the 1.9.6 version you already got a warning:
```
LaTeX Warning: Reference `mvBF_page_introduction' on page 1 undefined on input
line 3.
```
and apparently this gives a fatal error in the current master 1.9.7 (19420b7ae5fb764654df9d949d62429ae1432df2) due to the fix #9761 so looks like the problem was already present but never noticed.

It is hard to tell what caused the warning in the 1.9.6 version in the first place (did doxygen give a warning regarding `mvBF_page_introduction` ?
There is probably a source file that refers to `mvBF_page_introduction` (`\ref`, \link\`, ...) but it is not possible to tell what happened.

To be able to find the problem:
- Can you please attach a, small, self contained example (source+configuration file in a, compressed, tar or zip file) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

(Note when referring to a version please give the information from `doxygen -v` and not just a date (might be multiple versions created on one day.)

**Comment by StefanBattmer** [comment]
Thanks @albert-github for pointing me to #9761! We will first dig deeper on our side then!

**Comment by albert-github** [comment]
Well I think this problem has 2 sides:
- your side fixing the problem
- doxygen side shouldn't crash, so when you found the problem on your side it might be "easily" possible to create a small example so we can see what we can do on the doxygen side as well to prevent the crash.

Did you get a doxygen warning about this that a reference could not be found?

**Comment by albert-github** [comment]
Good that the basic problem is solved (i.e. the crash) and when I understood it well there was a difference between your used *.sty file and the default doxygen.sty file.
- How do you use the style sheets as when the doxygen.sty from the release is present I think the problem should not occur.
- did you overwrite the `doxygen.sty` in the `latex` directory with your own file?
  - if so I think this is bad and you should have a look at `LATEX_EXTRA_STYLESHEET` and place in this file the commands you redefine (with `\renewcommand`) or commands you add or what else you need.

Regarding the `LaTeX Warning: Reference `mvBF_page_introduction' on page 1 undefined on input` it would me good to have a minimal example so we can try to find the underlying problem of this missing reference.

**Comment by StefanBattmer** [comment]
Dear @albert-github, I managed to generate a small example displaying the problem. It seems like each Doxygen \ref command results in a corresponding warning with the name of that section so if you have a `\ref MySection` on a page `MyPage` inside the Doxygen source code you will get a `LaTeX Warning: Reference `MyPage_MySection on page 1 undefined on` warning. The zip archive contains the example, the console output when calling _make_ in the PDF folder afterwards and the files that are generated then. We fiddled around with what was recommended in #9726 A LOT, but never managed to get our \note and \attention tags customized (see yellow and blue in the PDF from my example). I hope this makes sense.
[doxygenIssue9781.example.zip](https://github.com/doxygen/doxygen/files/10448410/doxygenIssue9781.example.zip)

**Comment by albert-github** [comment]
I think you are referring to the lines:
```
LaTeX Warning: Reference `ThePage_ThePage_subsection' on page 1 undefined on in
LaTeX Warning: Reference `mdf@pagelabel-1' on page 1 undefined on input line 8.
LaTeX Warning: Reference `mdf@pagelabel-2' on page 1 undefined on input line 11
```
but the console output is only partly of interest what we see there as well  is that pdflatex is called 4 times and that the warnings are only produced with the first pdflatex run.
LaTeX processors in general cannot get all the hyperlink references etc resolved in the first loop and therefor multiple pdflatex runs are necessary, in the make.bat / Makefile doxygen tries to accomplish this for you.
So only the last pdflatex run is, in respect to the output relevant, and this output ca be found in the refman.log.


I don't know the `mdframed` package but it looks like that the package is, in a maintenance sense, quite dead. Also the author looks like not to be active on GitHub or tex.stackexchange.
That said it looks like something of the package still works and according to the documentation can give nice results.


Note: for the `title.png` and `logo.png` why not use something like:
```
LATEX_EXTRA_FILES += title.png
LATEX_EXTRA_FILES += logo.png
```
in the doxygen settings file so the files are automatically copied.

**Comment by StefanBattmer** [comment]
@albert-github: Thanks for the advice! I make use of your suggestions regarding the `LATEX_EXTRA_FILES` now and understand that these warnings are probably nothing to worry about. However my remaining concern would be with the ability to adjust the PDF output with custom style data. From what I read here and also in the related issue you strongly suggest NOT to override the `doxygen.sty` file but use the `LATEX_EXTRA_STYLESHEET` instead to apply customization. We kind of do both :-( We took a Doxygen generated `doxygen.sty` applied customisation there and now feed this to the `LATEX_EXTRA_STYLESHEET` option. This is the only way we managed to get this working even though I would like to maintain only OUR custom stuff in some customization.sty file that is added to `LATEX_EXTRA_STYLESHEET`. This however we never managed to get working. Our stuff then is simply ignored. Any idea what we could do wrong there?

I would also happily get rid of the `mdframd` package. However we never got an alternative working. If you have a suggestion we would immediately give it a shot.

Again thanks for the nice support!

**Comment by albert-github** [comment]
> From what I read here and also in the related issue you strongly suggest NOT to override the doxygen.sty file but use the LATEX_EXTRA_STYLESHEET instead to apply customization. 

Indeed.

> We kind of do both :-( We took a Doxygen generated doxygen.sty applied customisation there and now feed this to the LATEX_EXTRA_STYLESHEET option

Doesn't look good, but I don't know what you did so having the customized `doxygen.sty` would be an advantage (together with the output of `doxygen -x` i.e. settings used by you different from the default settings)

`mdframed` replacement, I never used it so it is a bit hard to comment on this. 
Regarding a replacement I have no idea either and it will depend on the features you use. I assume you looked around already and googled "latex mdframed replacement" (and "latex mdframe replacement" as this also gives some hits). You could also try to ask a question on https://tex.stackexchange.com

**Comment by albert-github** [comment]
Some points I would start to do:

- update the `header.tex` and `footer.tex` to to current version (see `doxygen -w` how to obtain these default versions), so you can make use of the extra, automatic replacement, facilities that are incorporated, see the lines starting with `%%BEGIN` and  `%%END`, and also the automatic replacements with the `$` variables like `$extralatexstylesheet` and `$extralatexpackages` so that the following advise would work. Maybe you don't need parts of it anymore and can accomplish this by means of the defaults.
- set:
  ```
  EXTRA_PACKAGES = xcolor mdframed
  ```
- update only the required parts in your own stylesheet, this would look like:
  ```
  \definecolor{hinweis}{rgb}{0.9,0.96,1}
  \definecolor{achtung}{rgb}{1,0.97,0.85}

  \mdfdefinestyle{myhinweis}{roundcorner=2pt,leftmargin=5,rightmargin=5,backgroundcolor=hinweis}
  \mdfdefinestyle{myachtung}{roundcorner=2pt,leftmargin=5,rightmargin=5,backgroundcolor=achtung}

  \renewenvironment{DoxyAttention}[1]{%
    \begin{mdframed}[style=myachtung]%
      \begin{DoxyDesc}{#1}%
  }{%
      \end{DoxyDesc}%
    \end{mdframed}%
  }

  \renewenvironment{DoxyNote}[1]{%
    \begin{mdframed}[style=myhinweis]%
      \begin{DoxyDesc}{#1}%
  }{%
      \end{DoxyDesc}%
    \end{mdframed}%
  }

  ```
  and this file you can feed to `LATEX_EXTRA_STYLESHEET`

**Comment by StefanBattmer** [comment]
@albert-github: Hi again. I made some progress and now may have something more solid for your analysis. Please see the latest attachment. This can be used to reproduce the 2 remaining issues I am facing now:
* With `PDF_HYPERLINKS=YES` Doxygen seems to stumble over the number in the section title and tries to make a numbered list from it. This generates strange HTML output and later forces the LaTeX compiler to stop because apparently the generated tex code is invalid.
* With `PDF_HYPERLINKS=NO` the PDF is generated fine BUT the links are no longer generated. As expected page numbers are added to the document but NO links are generated. This is happening since I updated the header.tex file to the Doxygen version I am using (1.9.6). The zip archive also contains the old (header-old.tex) file from Doxygen 1.8.17. Using this one creates correct links for the page numbers BUT doesn't add colors to my attention and note sections.
[doxygenIssue9781.PDFLinksBroken.zip](https://github.com/doxygen/doxygen/files/10470543/doxygenIssue9781.PDFLinksBroken.zip)

**Comment by albert-github** [comment]
I think there are a number of issues here
- mixing an old version with new code `doxysectlink` / `doxysectref` which are not used in the 1.9.6 version (it is only in the master version)
- the `doxysectlink` for the master version are already in the official `doxygen.sty` for that version and when one wants to overrule them one has to use `\renewcommand`.
- I'm getting a problem regarding `\end{DoxyEnumerate}`, do you get this as well?
  Looks like there is some interaction between `\page` ,`tableofcontents` and `\ref` in the part:
  ```
  \page ThePage The Page

  \tableofcontents

  Check it here \ref ThePage_subsection
  ```
  I'll have to look into this.
- You write the `PDF_HYPERLINKS=NO` in your old version you got `link` that are gone now, with `PDF_HYPERLINKS=NO` there should not be any hyperlinks at all. So I don't get this.

**Comment by albert-github** [comment]
The construct `1.` in the section title is a bit unusual, but should not lead to problems.

I've just pushed a proposed patch, pull request #9805

**Comment by StefanBattmer** [comment]
My understanding was that when `PDF_HYPERLINKS=NO` you also get links but instead of the section title, the page number of the target link is added to the document instead and this can be clicked and gets you to the corresponding section then as well. At least this was the case with the old version. If that is incorrect or has been a bug that has been fixed I am happy to go with the text based hyperlinks. I was confused since the header-old.tex when included in the example does exactly that. The difference can be seen in the resulting refman.aux. I think the line making the difference is the one starting with `\newlabel{ThePage_ThePage_subsection}`. I will check both patches (this one and the one from #9779 on Monday.

Thanks again!

**Comment by albert-github** [comment]
I think it is a misunderstanding as  by default when `PDF_HYPERLINKS=NO` there should not be any hyperlink, would be strange to have things like (`p1`) and a text link in one document. the hyperlinked texts are intended for using the pdf and the `p1` for printing.
I can see in the `header-old.ex` some references to hyperlinks, probably it has been generated based on a Doxyfile with `PDF_HYPERLINKS=YES` and later on this is used with a setting of `PDF_HYPERLINKS=NO`.

With the `%%BEGIN ...`this all has been made a bit more flexible. With the fixes in master (especially #9761) there is more flexibility in how the links should look, but it is up to the user (`renewcommand....`) to accomplish his specific wishes / layout.

**Comment by albert-github** [comment]
The code regarding #9805 has been integrated in master on GitHub.

**Comment by StefanBattmer** [comment]
@albert-github: #9805 fixes the issue with the 'list items in section titles'! Regarding the discussion about the `PDF_HYPERLINKS=YES/NO` here: I am only using the following `doxygen.custom.sty` as an input for the `LATEX_EXTRA_STYLESHEET` now:

```
\RequirePackage{xcolor}
\RequirePackage{mdframed}

\definecolor{MyNoteColor}{rgb}{0.9,0.96,1}
\definecolor{MyAttentionColor}{rgb}{1,0.97,0.85}

\mdfdefinestyle{MyNote}{roundcorner=2pt,leftmargin=5,rightmargin=5,backgroundcolor=MyNoteColor}
\mdfdefinestyle{MyAttention}{roundcorner=2pt,leftmargin=5,rightmargin=5,backgroundcolor=MyAttentionColor}

\renewenvironment{DoxyAttention}[1]{%
  \begin{mdframed}[style=MyAttention]%
    \begin{DoxyDesc}{#1}%
}{%
    \end{DoxyDesc}%
  \end{mdframed}%
}

\renewenvironment{DoxyNote}[1]{%
  \begin{mdframed}[style=MyNote]%
    \begin{DoxyDesc}{#1}%
}{%
    \end{DoxyDesc}%
  \end{mdframed}%
}
```

to get the `\note` and `\attention` sections colorized. This works perfectly now without the need to do anything else. So I am a happy guy :-)

I still have one remaining question though: After your last comment I fiddled around with `doxygen -w` again to see if the optional parameter `[configFile]` indeed is meant to create different output based on the doxyfile fed into it and I get the same output for LaTeX whether or not `PDF_HYPERLINKS` is set to YES or NO. Is there a chance this did change sometime? Doxygen DOES behave like documented here(https://www.doxygen.nl/manual/config.html#cfg_pdf_hyperlinks) indeed so definitely no problem. The hyperlink embedded into the page number was/would be nice though but the way it is right now is just as perfect! Thanks a lot for your patience!

**Comment by albert-github** [comment]
In the current code I indeed don't see any reference to the `Doxyfile` in this case an in older versions I did see a difference when generating the "default" versions. So yes I think it did change in the mean time.
The specific "default" versions based on the settings is not necessary anymore now due to the handling by means of the `%%BEGIN...` (looks like since 1.9.2).
The `[Doxyfile]` argument could be removed, but is left for the sake of compatibility (although it doesn't do anything as far as I can see).
